### PR TITLE
Patch readline-import issue which causes VLLMGenerator hang

### PR DESCRIPTION
**Pull request by pzhan9**
The reason is explained in the in-line comments. This issue seems to be caused by a [recent change in vllm](https://github.com/vllm-project/vllm/pull/46718), which introduces tilelang import, and subsequently triggers the import chain.

Note that this is not a monarch bug. Although one can argue if monarch spawns the ProcMesh in a new session, [i.e. change `setpgid` to `setsid`](https://github.com/meta-pytorch/monarch/blob/a4af2fe08e8fd7fceac736089631bacf0baf5c7b/hyperactor_mesh/src/proc_launcher/native.rs#L355), readline-import's behavior will be tolerant, but the Monarch team does not think simply tolerating such behavior justifies an critical infra change on the Monarch side. More importantly, the Monarch side is not sure whether it might require sharing session in the future, and it does not want to lose that option too casually. As a result, the Monarch prefers to workaround it on the user side, especially given the workaround is reasonably small. cc: @shayne-fletcher

In addition, add `torchvision` to README's installation instruction since it is required by vllm recently.

### CODE REVIEW

File: torchtitan/experiments/rl/README.md [hunk 1]
```diff
@@ -48,10 +48,13 @@ uv pip install flash-attn-3 --extra-index-url=https://download.pytorch.org/whl/t
 uv pip install --no-deps "git+https://github.com/thinking-machines-lab/batch_invariant_ops.git@main"
 ```
 
-4. Install PyTorch nightly, pre-built vllm wheel (based on PyTorch nightly version), and torchcomms nightly.
+4. Install PyTorch and torchvision nightlies, pre-built vllm wheel (based on PyTorch nightly version), and torchcomms nightly.
+
+`torchvision` is only needed because the current vllm nightly imports it during kernel warmup; TorchTitan RL does not otherwise require it.
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/README.md:53]: why wouldn't vllm nightly wheel already depend on this and automatically install?

File: torchtitan/experiments/rl/train.py [hunk 2]
```diff
@@ -68,12 +68,43 @@ def breakable_cudagraph_env(generator_cfg) -> dict[str, str]:
...
 
+def _bootstrap_generator() -> None:
+    """``bootstrap`` setup callable for VLLMGenerator."""
+    import os
+
+    # VLLMGenerator encounters the following issue:
+    # 1. vLLM's jit_monitor transitively imports readline through the following chain:
+    #      vLLM -> tilelang  -> tvm  -> tvm's base.py -> import readline
+    #      https://github.com/vllm-project/vllm/blob/b23bd73f540175f9e117eaee5029cd7d8df63964/vllm/utils/jit_monitor.py#L451
+    #      https://github.com/tile-ai/tvm/blob/28a0d34420d2fa9bc71fc891445a3f1396fca759/python/tvm/base.py#L71
```
Comment (wwwjn) [inline torchtitan/experiments/rl/train.py:86]: My read is monarch is not compatiable with vllm. It should be either fixed in monarch or vllm, not user side? And this patch is hacking and hard to understand from RL infra user level

File: torchtitan/experiments/rl/train.py [hunk 3]
```diff
@@ -152,9 +183,10 @@ def _spawn_proc_mesh(
     role_gpus_per_node = role_world_size // nodes
     provisioner = PerHostProvisioner(total_gpus=gpus_per_node)
     env = provisioner.allocate(role_gpus_per_node, extra_env=extra_env)
+    bootstrap = _bootstrap_generator if role == "generator" else _preimport_torch
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/train.py]: should be passed in as an arg of `_spawn_proc_mesh`, instead of if-else here? This function shouldn't need to know such details

File: torchtitan/experiments/rl/train.py (hunk 2 above)
Comment (pzhan9) [reply torchtitan/experiments/rl/train.py:86]: The culprit is `readline`. For some versions' `readline`, you cannot import it in a background process. This is an old problem. The oldest one I found was in 2012: https://github.com/python/cpython/issues/59097 

It also bit `torch.distrubuted` recently: https://github.com/pytorch/pytorch/issues/159645. In that issue, the user reported that they could not run `torch.distributed` in a background process. This issue is very similar to what we saw here: `vllm` does `import readline` unintentionally, which blocks `vllm` from being used in a background process.

Monarch is being caught in the crossfire, because it uses background processes to run its actors. But using background process is not wrong. That is why we do not want to make a big change to Monarch just to cater this decade old issue.

Can we patch it in vllm, or maybe its dependencies, e.g. tvm? It is possible. I have not looked into that. But even if that is technically possible, it might take a while to resolve, so we still need some patch here to unblock.

File: torchtitan/experiments/rl/README.md (hunk 1 above)
Comment (pzhan9) [reply torchtitan/experiments/rl/README.md:53]: vllm is installed from the wheel, and torchvision is not in the wheel's `Requires-Dist`. Probably that is because vllm excludes it here (disclaim: I have not checked e2e):

https://github.com/vllm-project/vllm/blob/9dd62d80abad5d2ebeb0a4a60ad005a23c23d0e9/use_existing_torch.py#L44

File: torchtitan/experiments/rl/train.py (hunk 2 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/train.py:86]: I understand it's easier to fix from user side, but I still think it should not be a user side fix. I guess a more clear fix path is landing this temporary fix, and raise an github issue in tvm (even a direct PR in tvm)

> background process is not wrong

Choosing background process as the design for monarch sounds correct, but the error is originally caused by vllm upstream. Just want to fix at the right place

File: torchtitan/experiments/rl/train.py [hunk 4]
```diff
@@ -68,12 +69,43 @@ def breakable_cudagraph_env(generator_cfg) -> dict[str, str]:
...
+    # 2. Monarch spawns procs in a background process group.
+    #
+    # As a result, when spawning the VLLMGenerator mesh, readline-import touches the
+    # controlling terminal in background processes, and kernel sends SIGTTOU, which stops
+    # the process.
+    #
+    # Strictly speaking this should be a tvm bug, since it should not assume the import
+    # happens in a foreground process. To workaround it, we do a proactive readline-import
+    # warmup here with SIGTTOU being temporarily blocked. Then the subsequent readline-import
+    # in tvm will not do tcsetattr again since it is already done here.
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/train.py:97]: let's add a TODO and follow what @wwwjn suggests if that makes sense

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/experiments/rl/train.py (hunk 2 above)
Comment (pzhan9) [reply torchtitan/experiments/rl/train.py:86]: I submitted a PR and opened an issue as well in case they prefer a different fix:

https://github.com/tile-ai/tvm/issues/55
https://github.com/tile-ai/tvm/pull/54
